### PR TITLE
Qi UTXO-set PMT with header commitment

### PR DESCRIPTION
**Pull request by jdowning100**
Todo:
 - Completely remove setCurrentUTXOSet() and db accessor methods from codebase
 
 
 Tested with Quai transactions and Qi transactions. Respective state roots are updated and blockchain progresses.

### CODE REVIEW

File: consensus/blake3pow/consensus.go [hunk 1]
```diff
@@ -451,7 +451,7 @@ func (blake3pow *Blake3pow) Finalize(chain consensus.ChainHeaderReader, header *
 			}
 		}
 	}
-
+	header.SetQiRoot(state.UTXORoot())
 	header.SetRoot(state.IntermediateRoot(true))
```
Comment (wizeguyy) [inline consensus/blake3pow/consensus.go:455]: To avoid ambiguity, please rename the old "root" logic and methods to be `QuaiRoot` or `EVMRoot`. The renames could/should be their own commit, to make it easy to review

File: consensus/blake3pow/consensus.go [hunk 2]
```diff
@@ -451,7 +451,7 @@ func (blake3pow *Blake3pow) Finalize(chain consensus.ChainHeaderReader, header *
 			}
 		}
 	}
-
+	header.SetQiRoot(state.UTXORoot())
```
Comment (wizeguyy) [inline consensus/blake3pow/consensus.go:454]: Here you call it QiRoot, elsewhere its called UTXORoot. I think lets just call it UTXORoot, and that will work nicely when we rename the old logic to EVMRoot.

Then one day if we reorganize the tree, we may actually split the EVM root into an AccountsRoot and ContractsRoot, or something like that

File: consensus/blake3pow/consensus.go (hunk 1 above)
Comment (wizeguyy) [reply consensus/blake3pow/consensus.go:455]: e.g. in the next file, this:
```
statedb, err := state.New(parent.Root(), parent.QiRoot(), ...)
```
would be clearer as

```
statedb, err := state.New(parent.EvmRoot(), parent.UtxoRoot(), ...)
```

File: core/headerchain.go [hunk 3]
```diff
@@ -430,9 +430,9 @@ func (hc *HeaderChain) SetCurrentState(head *types.Header) error {
 		return nil
 	}
 
-	if err := hc.setCurrentUTXOSet(head); err != nil {
+	/*if err := hc.setCurrentUTXOSet(head); err != nil {
```
Comment (wizeguyy) [inline core/headerchain.go:433]: ded coad

File: core/state/statedb.go [hunk 4]
```diff
@@ -524,6 +540,106 @@ func (s *StateDB) getStateObject(addr common.InternalAddress) *stateObject {
...
+	// Delete the utxo from the trie
+	if err := s.utxoTrie.TryDelete(rawdb.UtxoKey(hash, index)); err != nil {
+		s.setError(fmt.Errorf("deleteUTXO (%x) error: %v", hash, err))
+	}
+}
+
+// CreateUTXO explicitly creates a UTXO entry.
+func (s *StateDB) CreateUTXO(hash common.Hash, index uint32, utxo *types.UtxoEntry) {
+	//s.utxoObjects[hash] = utxo
+	data, err := rlp.EncodeToBytes(utxo)
```
Comment (wizeguyy) [inline core/state/statedb.go:609]: should be proto encode, no?

File: core/state/statedb.go [hunk 5]
```diff
@@ -524,6 +540,106 @@ func (s *StateDB) getStateObject(addr common.InternalAddress) *stateObject {
...
+	}
+	// Delete the utxo from the trie
+	if err := s.utxoTrie.TryDelete(rawdb.UtxoKey(hash, index)); err != nil {
+		s.setError(fmt.Errorf("deleteUTXO (%x) error: %v", hash, err))
+	}
+}
+
+// CreateUTXO explicitly creates a UTXO entry.
+func (s *StateDB) CreateUTXO(hash common.Hash, index uint32, utxo *types.UtxoEntry) {
+	//s.utxoObjects[hash] = utxo
```
Comment (wizeguyy) [inline core/state/statedb.go:608]: ded coad

File: core/state/statedb.go [hunk 6]
```diff
@@ -524,6 +540,106 @@ func (s *StateDB) getStateObject(addr common.InternalAddress) *stateObject {
...
+		defer func(start time.Time) { stateMetrics.WithLabelValues("AccountUpdates").Add(float64(time.Since(start))) }(time.Now())
+	}
+	// Delete the utxo from the trie
+	if err := s.utxoTrie.TryDelete(rawdb.UtxoKey(hash, index)); err != nil {
+		s.setError(fmt.Errorf("deleteUTXO (%x) error: %v", hash, err))
+	}
+}
+
+// CreateUTXO explicitly creates a UTXO entry.
+func (s *StateDB) CreateUTXO(hash common.Hash, index uint32, utxo *types.UtxoEntry) {
```
Comment (wizeguyy) [inline core/state/statedb.go:607]: Just for clarity, could you rename these input args. Maybe: `txHash`, `outIdx`, `utxo`?

File: core/state/statedb.go (hunk 4 above)
Comment (jdowning100) [reply core/state/statedb.go:609]: The state trie uses RLP encoding. I can change the encoding, but I'm not confident that the rest of the state trie would continue to work (such as snapshots)

File: consensus/blake3pow/consensus.go (hunk 1 above)
Comment (jdowning100) [reply consensus/blake3pow/consensus.go]: i have renamed the state root to EVMRoot

File: consensus/blake3pow/consensus.go (hunk 2 above)
Comment (jdowning100) [reply consensus/blake3pow/consensus.go]: I have renamed QiRoot to UTXORoot and Root to EVMRoot

File: core/state/statedb.go (hunk 6 above)
Comment (jdowning100) [reply core/state/statedb.go]: Done

File: core/state/statedb.go (hunk 5 above)
Comment (jdowning100) [reply core/state/statedb.go]: Removed

File: core/headerchain.go (hunk 3 above)
Comment (jdowning100) [reply core/headerchain.go]: removed

File: core/state/statedb.go (hunk 4 above)
Comment (gameofpointers) [reply core/state/statedb.go:581]: we still havent changed the state and evm to use protobuf, down the line tasks

File: core/rawdb/accessors_chain.go [hunk 7]
```diff
@@ -1531,7 +1531,7 @@ func WriteUtxo(db ethdb.KeyValueWriter, hash common.Hash, index uint32, utxo *ty
 
 func ReadUtxo(db ethdb.Reader, hash common.Hash, index uint32) *types.UtxoEntry {
 	// Try to look up the data in leveldb.
-	data, _ := db.Get(utxoKey(hash, index))
+	data, _ := db.Get(UtxoKey(hash, index))
```
Comment (gameofpointers) [inline core/rawdb/accessors_chain.go:1534]: can we change the UTXO database accessors to use the protobuf encoding? i think the db changes went in first and then the UTXO pr and we missed this, just have to make sure that there is no rlp import in the rawdb module

File: core/state_processor.go [hunk 8]
```diff
@@ -154,7 +167,12 @@ func NewStateProcessor(config *params.ChainConfig, hc *HeaderChain, engine conse
 	if sp.cacheConfig.SnapshotLimit > 0 {
 		// TODO: If the state is not available, enable snapshot recovery
 		head := hc.CurrentHeader()
-		sp.snaps, _ = snapshot.New(hc.headerDb, sp.stateCache.TrieDB(), sp.cacheConfig.SnapshotLimit, head.Root(), true, false)
+		sp.snaps, _ = snapshot.New(hc.headerDb, sp.stateCache.TrieDB(), sp.cacheConfig.SnapshotLimit, head.EVMRoot(), true, false)
+		/*var err error
```
Comment (gameofpointers) [inline core/state_processor.go:171]: comment? i saw that above the snaps were created for  UTXO, either we should have them or not, completely remove them, that refers to all the changes to snapshot module as well

File: params/config.go [hunk 9]
```diff
@@ -36,7 +36,7 @@ var (
 	Blake3PowColosseumGenesisHash  = common.HexToHash("0x40057f44be0809f939c7d70893d101abc74af1f1c535e406ec9909cfa2c20cbe")
 	Blake3PowGardenGenesisHash     = common.HexToHash("0x0a942bb7fa04b80d658a64d0ca49c62199503a796abc169f05a1863421a22098")
 	Blake3PowOrchardGenesisHash    = common.HexToHash("0x1f3743c323ec7a9a6a8150a33de81723abb70fa10ad49c35e365e48488430b56")
-	Blake3PowLocalGenesisHash      = common.HexToHash("0x31ac78ff6abcd99ad1fd8cd1b8ef54cfb28bdc72f7a38311e67a9e34fc2adb45")
+	Blake3PowLocalGenesisHash      = common.HexToHash("0x5c10b64d8a10ab138a5ef7abb009aebb441ddae5cad2446bbc140b1e0a0de81e")
```
Comment (gameofpointers) [inline params/config.go:39]: does it only change the local hash, we have to make changes to all the hashes for both consensus engines

File: core/state_processor.go (hunk 8 above)
Comment (jdowning100) [reply core/state_processor.go:171]: I will remove them

File: core/rawdb/accessors_chain.go (hunk 7 above)
Comment (jdowning100) [reply core/rawdb/accessors_chain.go:1534]: There should be no UTXO database accessors anymore. I removed them in the next PR: https://github.com/dominant-strategies/go-quai-libp2p/pull/111

File: params/config.go (hunk 9 above)
Comment (jdowning100) [reply params/config.go:39]: I'm not sure what changes you want to make here

File: core/state_processor.go (hunk 8 above)
Comment (jdowning100) [reply core/state_processor.go]: Removed

File: params/config.go (hunk 9 above)
Comment (wizeguyy) [reply params/config.go:39]: all of the hashes should be updated, right?

**Review by wizeguyy — DISMISSED** [review]

File: core/slice.go [hunk 10]
```diff
@@ -1608,7 +1609,8 @@ func (sl *Slice) cleanCacheAndDatabaseTillBlock(hash common.Hash) {
 
 	// Recover the snaps
 	if nodeCtx == common.ZONE_CTX && sl.ProcessingState() {
-		sl.hc.bc.processor.snaps, _ = snapshot.New(sl.sliceDb, sl.hc.bc.processor.stateCache.TrieDB(), sl.hc.bc.processor.cacheConfig.SnapshotLimit, currentHeader.Root(), true, true)
+		sl.hc.bc.processor.snaps, _ = snapshot.New(sl.sliceDb, sl.hc.bc.processor.stateCache.TrieDB(), sl.hc.bc.processor.cacheConfig.SnapshotLimit, currentHeader.EVMRoot(), true, true)
+		sl.hc.bc.processor.utxoSnaps, _ = snapshot.New(sl.sliceDb, sl.hc.bc.processor.utxoCache.TrieDB(), sl.hc.bc.processor.cacheConfig.SnapshotLimit, currentHeader.UTXORoot(), true, true)
```
Comment (gameofpointers) [inline core/slice.go:1613]: should we have snaps or not have them? if we do we should have full functionality

Comment (jdowning100) [reply core/slice.go]: removed

**Review by gameofpointers — APPROVED** [review]
